holopy/inference/tests/test_prior.py

```diff
@@ -17,60 +17,320 @@
-    assert_obj_close(p1.imag, Gaussian(2,1))
-    assert_equal(p1.guess, 1+2j)
-    assert_obj_close(p1.lnprob(0.5+1j), gold_sigma - np.log(2))
-    p2 = prior.ComplexPrior(1, Gaussian(2,1))
-    assert_equal(p2.guess, 1+2j)
-    p3 = prior.ComplexPrior(Uniform(0,2), 2)
-    assert_equal(p3.guess, 1+2j)
+
+GOLD_SIGMA = -1.4189385332 #log(sqrt(0.5/pi))-1/2
+
+class TestBasics(unittest.TestCase):
+    @attr("fast")
+    def test_cannot_instantiate_baseclass(self):
+        self.assertRaises(NotImplementedError, Prior)
+
+class TestUniform(unittest.TestCase):
+    @attr("fast")
+    def test_construction(self):
+        parameters = {'lower_bound':1, 'upper_bound':3, 'guess':2, 'name':'a'}
+        u = Uniform(*parameters.values())
+        self.assertTrue(isinstance(u, Prior))
+        for key, val in parameters.items():
+            self.assertEqual(getattr(u, key), val)
+
+    @attr("fast")
+    def test_upper_bound_larger_than_lower_bound(self):
+        self.assertRaises(ParameterSpecificationError, Uniform, 1, 0)
+        self.assertRaises(ParameterSpecificationError, Uniform, 1, 1)
+
+    @attr("fast")
+    def test_guess_must_be_in_interval(self):
+        self.assertRaises(ParameterSpecificationError, Uniform, 0, 1, 2)
+        self.assertRaises(ParameterSpecificationError, Uniform, 0, 1, -1)
+
+    @attr("fast")
+    def test_interval_calculation(self):
+        bounds = np.random.rand(2) + np.array([0,1])
+        u = Uniform(*bounds)
+        self.assertEqual(u.interval, np.diff(bounds))
+
+    @attr("fast")
+    def test_interval_is_property(self):
+        bounds = np.random.rand(2) + np.array([0,1])
+        u = Uniform(*bounds)
+        self.assertRaises(AttributeError, setattr, u, 'interval', 2)
+
+    @attr("fast")
+    def test_prob(self):
+        bounds = np.random.rand(2) + np.array([0,1])
+        u = Uniform(*bounds)
+        self.assertEqual(u.prob(0), 0)
+        self.assertEqual(u.prob(2), 0)
+        self.assertEqual(u.prob(1), 1/np.diff(bounds))
+
+    @attr("fast")
+    def test_lnprob(self):
+        bounds = np.random.rand(2) + np.array([0,1])
+        u = Uniform(*bounds)
+        self.assertEqual(u.lnprob(0), -np.inf)
+        self.assertEqual(u.lnprob(2), -np.inf)
+        self.assertTrue(np.allclose(u.lnprob(1), -np.log(np.diff(bounds))))
+
+    @attr("fast")
+    def test_sample_shape(self):
+        n_samples = 7
+        bounds = np.random.rand(2) + np.array([0,1])
+        u = Uniform(*bounds)
+        samples = u.sample(n_samples)
+        self.assertEqual(samples.shape, np.array(n_samples))
+        self.assertTrue(np.all(samples > bounds[0]))
+        self.assertTrue(np.all(samples < bounds[1]))
+
+    @attr("medium")
+    def test_sample_distribution(self):
+        n_samples = 100000
+        n_quantiles = 10
+        bounds = np.random.rand(2) + np.array([0,1])
+        u = Uniform(*bounds)
+        samples = u.sample(n_samples)
+        quantiles = np.quantile(samples, np.linspace(0, 1, n_quantiles))
```

Comment:
This line (104) fails on Travis because ```np.quantile```wasn't implemented until numpy version 1.15. I'd suggest using ```np.percentile```, although then you have to change it to
```python
         quantiles = np.percentile(samples, np.linspace(0, 100, n_quantiles))
```

Also, I have some other comments on the test:
* I would set a random seed, since a probabilistic test should fail with a small probability, you don't want to run the test suite 1000 times and have it fail once and be super confused as to why that's the case.
* 100,000 samples is a lot of samples. I'd make the number much fewer (e.g. 10,000 or 1,000) and the corresponding tolerance much smaller.
* Since we're physicists who like inference here, I'll suggest something a little more technical. By testing quantiles, you're only testing the distribution's value at 10 points. You can instead test the distribution at *every* value using what's called a [Kolmogorov-Smirnov test](https://en.wikipedia.org/wiki/Kolmogorov%E2%80%93Smirnov_test) -- there are other tests for equality of distribution, but I think this is the best test for the uniform distribution case. An advantage of a KS test here is that not only does it test for equality of distribution everywhere (rather than at 10 points), it is also *much* more efficient and very easy to calculate P-values for. Fortunately there is a KS test implemented in scipy (```scipy.stats.kstest```) which calculates P-values etc, so this is a easy change. I'm happy to chat if you have questions.
* Also, since we're still all physicists here, I'd rather see ```atol``` and ```rtol``` to be something you'd expect from the statistical test, rather than an arbitrary number (e.g. a P-value or something).

Finally, at this stage I'd rather have many B+ tests than a few A+ tests, so I'm happy with only fixes that partially address these comments.

Reply:
Can we talk about this at some point? I have tried to implement but got confused...

Reply:
Yeah! Can I swing by at 4:45 today?